### PR TITLE
Add event log

### PR DESCRIPTION
**Pull request by sug0**
Add event log from ADR 075.

Essentially, this is just a backport of PR https://github.com/tendermint/tendermint/pull/7825

### CODE REVIEW

**Comment by sug0** [comment]
@batconjurer it's alright, any review helps a bit. maybe you could have spotted something that I might have missed, since between this patchset and the original one from upstream, I've changed a few things.

**Comment by james-chf** [comment]
There is also this backport - https://github.com/tendermint/tendermint/pull/9470 (https://github.com/tendermint/tendermint/tree/feature/adr075-backport)

**Comment by sug0** [comment]
> There is also this backport (for v0.37-ish code I think) - [tendermint#9470](https://github.com/tendermint/tendermint/pull/9470) (https://github.com/tendermint/tendermint/tree/feature/adr075-backport)

the patchset I applied from the original PR is pretty much identical to this one, if not identical. I did change a few things for this specific PR though, like relocating the go package to a new place, and changing some reprs to adapt to 0.37.x data structures.

File: statesync/mocks/state_provider.go [hunk 1]
```diff
@@ -6,6 +6,7 @@ import (
 	context "context"
 
 	mock "github.com/stretchr/testify/mock"
+
```
Comment (james-chf) [inline statesync/mocks/state_provider.go:9]: This format change thing is a bit annoying, I got it as well when running `make mockery` then `make format` - ideally these files shouldn't be changing after they generate. I'm guessing we need to fix a specific version of `goimports` or smth, will look at upstream

**Review by james-chf — APPROVED** [review]
We need to update some mocks I think (also not

**Pull request merged**